###Assignment 5: Random Forest Team: White Walkers Student Names: Pratik Dhumal (SJSU ID: 012457185) Sourabh Namilikonda (SJSU ID: 012414948) Hrishikesh Rendalkar (SJSU ID: 011485994) Hemambujam Veeraraghavan (SJSU ID: 012482392) Business Objective: Being software engineers, we develop a number of mobile applications. We wanted to get an insight on the kind of application that are present in the market and work towards an application that would be suitable and useful for the consumers. Hence we decided to gather data on the the available applications in the play store, and perform data analysis on these applications and the user reviews.

##Dataset: The Playstore dataset we have chosen for this assignment holds data with regards to user ratings, reviews, size, number of installs and the price of the applications. This dataset had ample data, with about 4000 sample data, which served to be useful for the decision making algorithm.

##Data Wrangling: Using data from the above-mentioned dataset, we converted data from a plain text format to a list of lists in Python, that can be used for our assignment purpose.

##Data Interpretation: With the cleaned datset,We are trying to predict the decision of downloading an application from google play store which will be based on 4 factors ratings,reviews,no of downloads and size.

##Future Enhancements: In Future, we want to compare the apps from this playstore(Google) to other playstores(Apple) to get app popularity among different platforms and devices

In [68]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import random
from pyexcel_ods import get_data
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,classification_report
from IPython.display import Image  
from sklearn.externals.six import StringIO  
from sklearn.tree import export_graphviz
import pydot
%matplotlib inline

In [69]:
df = pd.read_excel('playstore.xlsx')
dataFrame=df[['Rating','Reviews','Size','Installs']]
dataFrame['Rating'].fillna(0, inplace = True)
dataFrame

/home/pratik/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py:5430: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


Rating  Reviews                Size     Installs
0         4.1      159                 19M      10,000+
1         3.9      967                 14M     500,000+
2         4.7    87510                8.7M   5,000,000+
3         4.5   215644                 25M  50,000,000+
4         4.3      967                2.8M     100,000+
5         4.4      167                5.6M      50,000+
6         3.8      178                 19M      50,000+
7         4.1    36815                 29M   1,000,000+
8         4.4    13791                 33M   1,000,000+
9         4.7      121                3.1M      10,000+
10        4.4    13880                 28M   1,000,000+
11        4.4     8788                 12M   1,000,000+
12        4.2    44829                 20M  10,000,000+
13        4.6     4326                 21M     100,000+
14        4.4     1518                 37M     100,000+
15        3.2       55                2.7M       5,000+
16        4.7     3632                5.5M     500,000+
17        4.5       27                 17M      10,000+
18        4.3   194216                 39M   5,000,000+
19        4.6   224399                 31M  10,000,000+
20        4.0      450                 14M     100,000+
21        4.1      654                 12M     100,000+
22        4.7     7699                4.2M     500,000+
23        0.0       61                7.0M     100,000+
24        4.7      118                 23M      50,000+
25        4.8      192                6.0M      10,000+
26        4.7    20260                 25M     500,000+
27        4.1      203                6.1M     100,000+
28        3.9      136                4.6M      10,000+
29        4.1      223                4.2M     100,000+
...       ...      ...                 ...          ...
10811     0.0        4                3.9M         100+
10812     4.1       80                 13M       1,000+
10813     0.0       20                2.7M      10,000+
10814     4.0      785                 31M      50,000+
10815     4.2     5775                4.9M     500,000+
10816     0.0        2                6.8M         100+
10817     4.0      885                8.0M     100,000+
10818     0.0       96                1.5M      10,000+
10819     3.3       52                3.6M       5,000+
10820     5.0       22                8.6M       1,000+
10821     0.0        6                2.5M          50+
10822     0.0        0                3.1M          10+
10823     0.0        1                2.9M         100+
10824     0.0       67                 82M      10,000+
10825     0.0        7                7.7M         100+
10826     4.0    88486  Varies with device   5,000,000+
10827     4.2      117                 13M       5,000+
10828     3.4      291                 13M      10,000+
10829     4.6      603                7.4M      10,000+
10830     3.8      881                2.3M     100,000+
10831     0.0       38                9.8M       5,000+
10832     3.8     1195                582k     100,000+
10833     4.8       44                619k       1,000+
10834     4.0        7                2.6M         500+
10835     0.0        0                9.6M          10+
10836     4.5       38                 53M       5,000+
10837     5.0        4                3.6M         100+
10838     0.0        3                9.5M       1,000+
10839     4.5      114  Varies with device       1,000+
10840     4.5   398307                 19M  10,000,000+

[10841 rows x 4 columns]

In [70]:
#data purification
#data['Rating'].fillna(0, inplace = True)


def purify_rating(X):
    for ind,s in enumerate(X):
        if s>5:
            print(ind, s)
            X[ind]=s/10
    return X

def purify_price(s):
    s=s.replace("Everyone","0")
    return s.replace("$","")

def purify_installs(s):
    #print(type(s) )
    s=s.replace("+","")
    s=s.replace("Free","0")
    return s.replace(",","")

def purify_size(s):
    #print(type(s) )
    s=s.replace(",","")
    s=s.replace("+","")
    s=s.replace("M","")
    s=s.replace("k","")
    return s.replace("Varies with device","0")

In [71]:
dataFrame['Rating']=purify_rating(dataFrame['Rating'])
dataFrame['Rating']= list(map(float, dataFrame['Rating']))
dataFrame['Installs'] = [purify_installs(str(s)) for s in dataFrame['Installs']]
dataFrame['Installs'] = list(map(float, dataFrame['Installs']))
dataFrame['Size'] = [purify_size(str(s)) for s in dataFrame['Size']]
dataFrame['Size'] = list(map(float, dataFrame['Size']))

/home/pratik/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/home/pratik/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/home/pratik/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the docum

In [72]:
def addNewColumnForBoolean(installs,size,reviews,rating):
    if installs > 10000 and size < 30 and reviews > 1500 and rating > 4.4:
        val = 'Yes'
    else:
        val = 'No'
    return val

In [73]:
#dataFrame['Probability_Downloading'] =[addNewColumnForBoolean(installs,size) for installs in dataFrame['Installs'] ]
#for x in range(0, 10840):
#    installs=dataFrame['Installs'];
#    size=dataFrame['Size'];
#    reviews=dataFrame['Reviews'];
#    rating=dataFrame['Rating'];
#    dataFrame['Probability_Downloading'] =addNewColumnForBoolean(installs,size,reviews,rating)
list_value=[]
for index, row in dataFrame.iterrows():
    list_value.append(addNewColumnForBoolean(row['Installs'],row['Size'],row['Reviews'],row['Rating']))
   
    
dataFrame['Probability_Downloading']=list_value
x=dataFrame

/home/pratik/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  del sys.path[0]


In [106]:
y=dataFrame['Probability_Downloading']
#adding some randomness to the dataset to make it more realistic
for i in range(0,5000):
     if(i%2==0):
            dataFrame.loc[i,'Probability_Downloading']='Yes'
             
 
           
        
       
        
         
    
     
             
        
       
   # print(str(index) +" "+ row['Probability_Downloading']+" "+y)
y=dataFrame['Probability_Downloading'] 

/home/pratik/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [107]:
x=dataFrame.drop('Probability_Downloading',axis=1)
x

Rating  Reviews   Size    Installs
0         4.1      159   19.0     10000.0
1         3.9      967   14.0    500000.0
2         4.7    87510    8.7   5000000.0
3         4.5   215644   25.0  50000000.0
4         4.3      967    2.8    100000.0
5         4.4      167    5.6     50000.0
6         3.8      178   19.0     50000.0
7         4.1    36815   29.0   1000000.0
8         4.4    13791   33.0   1000000.0
9         4.7      121    3.1     10000.0
10        4.4    13880   28.0   1000000.0
11        4.4     8788   12.0   1000000.0
12        4.2    44829   20.0  10000000.0
13        4.6     4326   21.0    100000.0
14        4.4     1518   37.0    100000.0
15        3.2       55    2.7      5000.0
16        4.7     3632    5.5    500000.0
17        4.5       27   17.0     10000.0
18        4.3   194216   39.0   5000000.0
19        4.6   224399   31.0  10000000.0
20        4.0      450   14.0    100000.0
21        4.1      654   12.0    100000.0
22        4.7     7699    4.2    500000.0
23        0.0       61    7.0    100000.0
24        4.7      118   23.0     50000.0
25        4.8      192    6.0     10000.0
26        4.7    20260   25.0    500000.0
27        4.1      203    6.1    100000.0
28        3.9      136    4.6     10000.0
29        4.1      223    4.2    100000.0
...       ...      ...    ...         ...
10811     0.0        4    3.9       100.0
10812     4.1       80   13.0      1000.0
10813     0.0       20    2.7     10000.0
10814     4.0      785   31.0     50000.0
10815     4.2     5775    4.9    500000.0
10816     0.0        2    6.8       100.0
10817     4.0      885    8.0    100000.0
10818     0.0       96    1.5     10000.0
10819     3.3       52    3.6      5000.0
10820     5.0       22    8.6      1000.0
10821     0.0        6    2.5        50.0
10822     0.0        0    3.1        10.0
10823     0.0        1    2.9       100.0
10824     0.0       67   82.0     10000.0
10825     0.0        7    7.7       100.0
10826     4.0    88486    0.0   5000000.0
10827     4.2      117   13.0      5000.0
10828     3.4      291   13.0     10000.0
10829     4.6      603    7.4     10000.0
10830     3.8      881    2.3    100000.0
10831     0.0       38    9.8      5000.0
10832     3.8     1195  582.0    100000.0
10833     4.8       44  619.0      1000.0
10834     4.0        7    2.6       500.0
10835     0.0        0    9.6        10.0
10836     4.5       38   53.0      5000.0
10837     5.0        4    3.6       100.0
10838     0.0        3    9.5      1000.0
10839     4.5      114    0.0      1000.0
10840     4.5   398307   19.0  10000000.0

[10841 rows x 4 columns]

In [100]:
rfc=RandomForestClassifier(n_estimators=100)

In [101]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3)

In [102]:
rfc.fit(x_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [103]:
rfc_pred=rfc.predict(x_test)

In [104]:
print(confusion_matrix(y_test,rfc_pred))
print('\n')
print(classification_report(y_test,rfc_pred))

[[2091  209]
 [ 445  508]]


             precision    recall  f1-score   support

         No       0.82      0.91      0.86      2300
        Yes       0.71      0.53      0.61       953

avg / total       0.79      0.80      0.79      3253



In [105]:
from sklearn.metrics import accuracy_score
y_pred=rfc.predict(x_test)
print(accuracy_score(y_test,y_pred))

0.7989548109437442


#When we check the same data set with same randomness in those value Random Forest give us better accuracy score as compared to decision tree. For decision tree we tried replacing the same randomness in previous assignment and accuracy was 0.73% but in Random Forest its 0.79 which is definately more as compared to decison tree.